rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,108 @@
+subsequent stage, there is a chance that oneTBB workers will interfere with OpenMP threads.
+This interference might result in slight oversubscription,
+which in turn might lead to underperformance.
+
+This problem can be resolved with an API that indicates when parallel computation is done,
+allowing worker threads to be released from the arena,
+essentially overriding the default block-time.<br>
+
+This problem can be considered from another angle. Essentially, if the user can indicate where
+parallel computation ends, they can also indicate where they start.
+
+<img src="parallel_block_introduction.png" width=800>
+
+With this approach, the user not only releases threads when necessary
+but also specifies a programmable block where worker threads should stick to the
+executing arena.
+
+## Proposal
+
+Let's consider the guarantees that an API for explicit parallel blocks can provides:
+* Start of parallel block:
+  * Indicates the point from which the scheduler can use a hint and stick threads to the arena.
+  * Serve as a warm-up hint to the scheduler, making some worker threads immediately available
+    at the start of the real computatin.
+* "Parallel block" itself:
+  * Scheduler can implement different busy-wait policies to retain threads in the arena.
+* End of parallel block:
+  * Indicates the point from which the scheduler can drop a hint
+    and unstick threads from the arena.
+  * Indicates that worker threads should ignore
+    the default block time (introduced by PR#1352) and leave.
+
+Start of parallel block:<br>
+The warm-up hint should have similar guarantees as `task_arena::enqueue` from a signal standpoint.
+Users should expect the scheduler will do its best to make some threads available in the arena.
+
+"Parallel block" itself:<br>
+The guarantee for retaining threads is a hint to the scheduler;
+thus, no real guarantee is provided. The scheduler can ignore the hint and
+move threads to another arena or to sleep if conditions are met.
+
+End of parallel block:<br>
+It can indicate that worker threads should ignore the default block time but
+if work was submitted immediately after the end of the parallel block,
+the default block time will be restored.
+
+But what if user would like to disable default block time entirely?<br>
+Because the heuristic of extended block time is unsuitable for the task submitted
+in unpredictable pattern and duration. In this case, there should be an API to disable
+the default block time in the arena entirely.
+
+```cpp
+class task_arena {
+    void indicate_start_of_parallel_block(bool do_warmup = false);
+    void indicate_end_of_parallel_block(bool disable_default_block_time = false);
+    void disable_default_block_time();
+    void enable_default_block_time();
+};
+
+namespace this_task_arena {
+    void indicate_start_of_parallel_block(bool do_warmup = false);
+    void indicate_end_of_parallel_block(bool disable_default_block_time = false);
+    void disable_default_block_time();
```

Comment:
The end-user doesn't know what the default block time is, and it will be platform dependent. The first set of functions indicate of region of interest, these `default_block_time` functions change a property on the task_arena that is not tied to a region. That makes me think it is better as a constraint. Are there known cases where this needs to be disabled then reenabled dynamically?  

If the first two functions became something like `retain_threads` and `release_threads`, what would these be named? What about `set_sleep_policy( sleep_quickly | sleep_slowly )` or something like that.

Reply:
Perhaps, it is a good idea to move it constraints. If you need different guarantees just use different arenas the same is  applicable to priorities.
If we include the property as part of the constraints which in turn represent HW Resources perhaps the name should represent how these resources will be used, like `greedy` or something.

@akukanov what do you think?

Reply:
A couple questions:

- How would users think about "greedy" relative to per-arena priorities? It might imply some kind of priority between a greedy normal arena and non-greedy normal arena, even though that wouldn't be the case.
- In suggesting `sleep_quickly` and `sleep_slowly`, which I admit are not great names, I was trying to find something that indicated more about the wastefulness of holding onto resources once you have them and *while there's nothing better to do with them* in contrast to a greediness in acquiring resources, perhaps from some other competing arenas. I think this is the key point of `disable|enable_default_block_time` -- while it is a form of greediness, it is more about the amount of wastefulness tolerated to reduce startup latency on the next parallel algorithm.

Reply:
When I am thinking about this, thoughts about being nice/responsive to the demand from other arenas appear in my head. But I am not sure how to better combine these two sets of API as they kind of mutually exclusive to each other. Consider, _expect more parallel work to appear but be responsive to resources demand from other arenas_. Actually, this counterintuition applies to the proposed design as well. Perhaps, we need to express that mutual exclusiveness somehow in the API.

Reply:
> expect more parallel work to appear but be responsive to resources demand from other arenas.

What part of the proposal is stating this?

Reply:
It is not stating explicitly, but sort of implying the question _What will it mean if I invoke `indicate_start_of_parallel_block` and call `disable_default_block_time` right after that?_